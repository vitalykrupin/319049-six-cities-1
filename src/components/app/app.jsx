import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Switch, Route, withRouter} from 'react-router-dom';
import {getAuthorizationStatus, getUser, getPendingAuthStatus} from '../../reducers/user/selectors';
import Header from '../header/header.jsx/index.js';
import PageWrapper from '../page-wrapper/page-wrapper.jsx';
import MainPage from '../main-page/main-page.jsx';
import Favorites from '../favorites/favorites.jsx';
import SignIn from '../sign-in/sign-in.jsx';
import OfferDetails from '../offer-details/offer-details.jsx';
import {ROUTES} from '../../constants/constants';
import withRedirectRoute from '../../hocs/with-redirect-route/with-redirect-route';
import ErrorMessage from '../error-message/error-message';
import NotFound from '../not-found/not-found';

const App = (props) => {
  const {pendingAuthorization, isAuthenticated, user} = props;

  return (
    <React.Fragment>
      {pendingAuthorization ? <div>Loading...</div> : (
        <PageWrapper location={props.location.pathname}>
          <Header
            isAuthenticated={isAuthenticated}
            user={user}
          />
          <Switch>
            <Route
              path={ROUTES.HOME}
              component={MainPage}
              exact
            />
            <Route
              path={ROUTES.LOGIN}
              component={withRedirectRoute(SignIn, isAuthenticated, ROUTES.HOME, true)}
            />
            <Route
              path={`${ROUTES.OFFER}/:id`}
              component={OfferDetails}
            />
            <Route
              path={ROUTES.FAVORITES}
              component={withRedirectRoute(Favorites, isAuthenticated, ROUTES.LOGIN)}
            />
            <Route
              path={ROUTES.ERROR}
              component={withRedirectRoute(ErrorMessage, isAuthenticated, ROUTES.HOME, true)}
            />
            <Route
              component={NotFound}
            />
          </Switch>
        </PageWrapper>
      )}
    </React.Fragment>
  );
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  isAuthenticated: getAuthorizationStatus(state),
  user: getUser(state),
  pendingAuthorization: getPendingAuthStatus(state),
});

App.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  user: PropTypes.shape({
    id: PropTypes.number,
    email: PropTypes.string,
    name: PropTypes.string,
    [`avatar_url`]: PropTypes.string,
    [`is_pro`]: PropTypes.bool,
  }),
  pendingAuthorization: PropTypes.bool,
  location: PropTypes.any,
};

export {App};
export default withRouter(connect(mapStateToProps)(App));
