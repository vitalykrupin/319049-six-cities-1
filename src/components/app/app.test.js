import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {App} from './app.jsx';

configure({adapter: new Adapter()});

const userMock = {
  avatarUrl: `/static/avatar/1.jpg`,
  email: `me@me.ru`,
  id: 1,
  isPro: false,
  name: `vitaly`,
};

describe(`App`, () => {
  it(`renders correctly`, () => {
    const app = shallow(
        <App
          location={`/`}
          isAuthenticated={true}
          cityOffers={[`Amsterdam`, `Paris`]}
          onCityClick={jest.fn()}
          pendingAuthorization={false}
          user={userMock}
        />
    );

    expect(shallowToJson(app)).toMatchSnapshot();
  });
});
