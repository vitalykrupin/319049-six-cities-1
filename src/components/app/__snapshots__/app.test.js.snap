// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<PageWrapper
  pageType="login"
>
  <AppHeader />
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/login"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/offer/:null"
      render={[Function]}
    />
  </Switch>
</PageWrapper>
`;
